src/pybind/mgr/dashboard/frontend/src/app/shared/services/notification.service.ts

```diff
@@ -123,10 +108,38 @@ export class NotificationService {
           application
         );
       }
+      this.queueToShow(config);
+    }, 10);
+  }
+
+  private queueToShow(config: CdNotificationConfig) {
+    this.cancel(this.queuedTimeoutId);
+    if (!this.queued.find((c) => _.isEqual(c, config))) {
+      this.queued.push(config);
+    }
+    this.queuedTimeoutId = window.setTimeout(() => {
+      this.showQueued();
+    }, 250);
+  }
+
+  private showQueued() {
+    const byTitle = {};
+    let n;
+    while ((n = this.queued.shift())) {
+      if (!byTitle[n.title]) {
+        byTitle[n.title] = [];
+      }
+      byTitle[n.title].push(n);
+    }
+    Object.values(byTitle).forEach((configs: CdNotificationConfig[]) => {
+      const config = configs[0];
+      if (configs.length > 1) {
+        config.message = '<ul>' + configs.map((c) => `<li>${c.message}</li>`).join('') + '</ul>';
```

Comment:
Just wondering if it'd make sense to let the calling component to unfold and format this array.

Reply:
Do you mean like optional doing that?
As notifications with descriptions are mostly used by error descriptions, I see no use case for a individual component, that usually only wants to throw one error at a time. This will mostly handle error messages from different components at the same time - like when the mgr crashes.

Reply:
Yeah, just optionally. Now I'm getting used to the Angular way, by separating logic (ts) from presentation (html, scss), I feel uncomfortable when I see layout/formatting within a service. I was suggesting that the the notification formatting fully happens within the notification component.

I think it currently uses `innerHTML` to inject the HTML right from this notification service, which I'm not sure is optimal, as this method will be run up to 12 times per change detection (unless `onPush` strategy is used, which I don't think is the case).

Reply:
The problem is that the notification component never calls `.show` only components that want to show a notification do. The notification component only subscribes to the notification service to get notified if a notification was shown.

`.shown` is called by many different components and they just want that their information is displayed in some organized fashion. IMHO this is an implementation detail all components don't have to care about, that's why it's in a private method, too ;)

You are referring to the method `renderTimeAndApplicationHtml`, which is also used to have an equal style for the toasty that is shown through the service and the notifications listing through the component in a drop down pop over.
But this method is pretty unrelated to the current changes, of course you a free to create an issue to craft it out in something like a `notification-style.service.ts`?